Acceptance tests

gemcfadyen opened the pull request:
Acceptance tests were written to drill out the logic for the Game class. Test coverage is showing as 100% for classes, except for Game.java, which is showing at 83% method coverage and 73% line coverage. Through Intellij I can see the untested code is all within the main method.

In places Spy's were used in order to capture behaviour from the Player and Prompt, used in the test assertions.

Other refactorings taken into account were:
- renaming UserPrompt to CommandPrompt as it's now used for both the Players and the Game
- introducing Player interface to allow for a PlayerSpy and a HumanPlayer to be used within the Game

gemcfadyen committed:
Acceptance test added to force players to take turns until the board is full

gemcfadyen committed:
Acceptance test added to force congratulations message to be printed when the board has a winning formation

gemcfadyen committed:
Acceptance test added to force message to be displayed when there is no winner at the end of the game

gemcfadyen committed:
Acceptanace added to force the final board to be printed by the console

gemcfadyen commented:
@jsuchy @ecomba - here is the work for my latest story - 100% code coverage outside the Game's main method, using hand rolled mocks. 

I've raised this PR to share/review with other apprentices. Feel free to also take a look if you are interested.

gemcfadyen commented:
good spot. Updated now. Thanks!

jsuchy commented on src/main/java/ttt/Game.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public void play() {
+        int index = PLAYER_ONE_INDEX;
```
Can you find a better name for this variable?

jsuchy commented on src/main/java/ttt/Game.java:
```diff
@@ -0,0 +1,66 @@
+    public void play() {
+        int index = PLAYER_ONE_INDEX;
+
+        while (board.hasFreeSpace()) {
```
Perhaps there's a better way to implement the control flow here. Can you find a way to remove the need for the break statement?

jsuchy commented on src/main/java/ttt/Game.java:
```diff
@@ -0,0 +1,66 @@
+        board.updateAt(nextMove, player.getSymbol());
+    }
+
+    private int switchPlayer(int index) {
```
same here, index isn't the best name.

jsuchy commented on src/main/java/ttt/HumanPlayer.java:
```diff
@@ -23,7 +24,8 @@ public int chooseNextMoveFrom(Board board) {
 
-    public PlayerSymbol getSymbols() {
+    @Override
+    public PlayerSymbol getSymbol() {
```
Guessing there's duplication amongst the different Player implementations, might want to consider abstract Player class instead of an interface. I don't usually recommend preferring inheritance...

jsuchy commented on src/test/java/ttt/GameTest.java:
```diff
@@ -0,0 +1,95 @@
+
+        game.play();
+
+        assertThat(gamePrompt.getLastMessagePrinted(), is("Congratulatory message printed"));
```
Can we ask the spy a different question? Do we care about the details of the message that's printed, or just that the prompt was told to print a particular message?

gemcfadyen replied:
I like this idea, as it makes the class a proper spy, rather than a mix between a spy and a stub returning canned answers.

ecomba commented on src/main/java/ttt/Game.java:
```diff
@@ -0,0 +1,66 @@
+            }
+            index = switchPlayer(index);
+        }
+        printBoardAndExitMessage();
```
The keyword `And` is usually a good indicator that a method is doing more than one thing. Is this the case in `printBoardAndExitMessage()`. If so, how could we change that keeping the code simple and clean?

jsuchy commented on src/test/java/ttt/PromptSpy.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public void printWinningMessage() {
+        lastMessagePrinted = "Congratulatory message printed";
```
At the very least, these "test" messages should be made into constants so that the tests aren't coupled to the details in here.

gemcfadyen committed:
Adresses pr comments: renaming variables to give more context, removing break clause from while loop, updating the PromptSpy to count how many times win/draw messages are printed

gemcfadyen committed:
Removes duplicate test case and aligns test names with the domain language

gemcfadyen commented:
Updated PR with the following changes based on the review comments:
- Renamed `index` variable to have more context
- Removed the `break` clause in the while loop by introducing a `hasWinner` flag
- Updated the PromptSpy to count how many times a win/draw message is printed and updated the corresponding tests
- Replaced the Player interface with an abstract Player class to prevent duplication of the `getSymbol` method and `Prompt` member variable
